src/lib/operators/sort.cpp

```diff
@@ -1,260 +1,283 @@
-            ++chunk_it;
-          }
-        }
+std::shared_ptr<AbstractOperator> Sort::_on_deep_copy(
+    const std::shared_ptr<AbstractOperator>& copied_input_left,
+    const std::shared_ptr<AbstractOperator>& copied_input_right) const {
+  return std::make_shared<Sort>(copied_input_left, _sort_definitions, _output_chunk_size);
+}
 
-        // Last segment has not been added
-        if (chunk_offset_out > 0u) {
-          auto value_segment = std::make_shared<ValueSegment<ColumnDataType>>(std::move(value_segment_value_vector),
-                                                                              std::move(value_segment_null_vector));
-          chunk_it->push_back(value_segment);
-        }
-      });
-    }
+void Sort::_on_set_parameters(const std::unordered_map<ParameterID, AllTypeVariant>& parameters) {}
 
-    for (auto& segments : output_segments_by_chunk) {
-      output->append_chunk(segments);
-    }
+std::shared_ptr<const Table> Sort::_on_execute() {
+  const auto& input_table = input_table_left();
+  for (const auto& column_sort_definition : _sort_definitions) {
+    Assert(column_sort_definition.column != INVALID_COLUMN_ID, "Sort: Invalid column in sort definition");
+    Assert(column_sort_definition.column < input_table->column_count(),
+           "Sort: Column ID is greater than table's column count");
+  }
 
-    return output;
+  std::shared_ptr<Table> sorted_table;
+
+  // Starting after the first (least significant) sort operation, this holds the order of the table as it has been
+  // determined so far. This is not a completely proper PosList on the input table as it might point to
+  // ReferenceSegments.
+  auto previously_sorted_pos_list = std::optional<PosList>{};
+
+  for (auto sort_step = static_cast<int64_t>(_sort_definitions.size() - 1); sort_step >= 0; --sort_step) {
+    const bool is_last_sorting_run = (sort_step == 0);
+
+    const auto& sort_definition = _sort_definitions[sort_step];
+    const auto data_type = input_table->column_data_type(sort_definition.column);
+
+    resolve_data_type(data_type, [&](auto type) {
+      using ColumnDataType = typename decltype(type)::type;
+
+      auto sort_impl = SortImpl<ColumnDataType>(input_table, sort_definition.column, sort_definition.order_by_mode);
+      previously_sorted_pos_list = sort_impl.sort(previously_sorted_pos_list);
+
+      if (is_last_sorting_run) {
+        // This is inside the for loop so that we do not have to resolve the type again
+        sorted_table = materialize_output_table(input_table, *previously_sorted_pos_list, _output_chunk_size);
+      }
+    });
   }
 
- protected:
-  const std::shared_ptr<const Table> _table_in;
-  const size_t _output_chunk_size;
-  const std::shared_ptr<std::vector<std::pair<RowID, SortColumnType>>> _row_id_value_vector;
-};
+  auto final_sort_definition = _sort_definitions[0];
+  // Set the ordered_by attribute of the output's chunks to the column the chunk was sorted by last (because this
```

Comment:
I could imagine removing the whole part in braces and replacing it somehow with above's "most significant".

Reply:
> Set the ordered_by attribute of the output's chunks according to the most significant sort operation, which is the column the table was sorted by last.

should be enough, don't you thnk?

Reply:
Yes.